### ISSUE TITLE
Memory leak WinUI 3

### ISSUE THREAD

**Issue by Omenrevlis** [Mapsui/Mapsui#2259]
**Mapsui Version**
4.1.1

**Mapsui Platform**
WinUI 3

**Device**
Windows 11

**Describe the bug**
When you create a Mapsui map either in code or by Xaml, when disposed it does not release all resources. This is true if you use 1 or 20. It's always the same resources. 

`Mapsui.IFeature[]	Count:1	Size(bytes):12
Mapsui.Styles.Color	Count:148	Size(bytes):3552
Mapsui.Animations.Easing	Count:13	Size(bytes):572
Action<Mapsui.Navigator>	Count:1	Size(bytes):32
Mapsui.Map+<>c	Count:1	Size(bytes):12
Mapsui.Animations.Easing+<>c	Count:1	Size(bytes):12
EmptyPartition<Mapsui.Animations.AnimationEntry<Mapsui.Viewport>>	Count:1	Size(bytes):12
Mapsui.Map[]	Count:1	Size(bytes):12
Mapsui.Tiling.Layers.TileLayer+<>c	Count:1	Size(bytes):12
EmptyPartition<Mapsui.Layers.ILayer>	Count:1	Size(bytes):12
Mapsui.Fetcher.FetchWorker[]	Count:1	Size(bytes):12
Mapsui.Styles.BitmapRegistry	Count:1	Size(bytes):396
			
			Total size (bytes)4648`

It's not a lot, but enough to keep a page in memory and if a new pages. With WinUI Navigation Template Studio template this can be brutal. It means any page that has a map UI memory is not released each time you visit the page. It will not be collected until the application stops. For some applications this could be a major memory leak issue. 

**To Reproduce**
1. In Visual Studio 2022 create a WinUI 3 project and page (you can use a blank page in Template Studio with a navigation control if you want to make it easy).  
2. Create two buttons in a stack panel and wire for their click events. 
3. create a class wide variable _Maps<Map>
4. In the first button new up a map, add two layers then add to _Maps collection. For fun you could add 20 or more. The results are the same with 1 or many.
5. in the second event loop through the collection and call dispose then _Maps.clear() to even remove the lists references. To make testing faster also call GC.Collect(); 
6. Take a snapshot and then open the snapshot and search for mapsui. You'll see the list I sent. If you don't trust the GC collected (it did, wait a while. Even an hour. Then take another snapshop. Those items stay in memory the whole time. 

**Expected behavior**
Expected behavior would be ZERO mapsui objects after calling Dispose and a garbage collection. 

**Screenshots**
![image](https://github.com/Mapsui/Mapsui/assets/150300234/ac4d5ef4-1368-4d22-8ff7-919bdd0ace37)

**Additional context**
Why not use System.Drawing.Color instead of specific ones for Mapsui?

**Comment by pauldendulk** [comment]
Thanks for your thorough analysis. Some initial reactions.

> Why not use System.Drawing.Color instead of specific ones for Mapsui?

System.Drawing.Color was not available on all platforms when we started, which was in the PocketPC/Silverlight days. It looks like it is an option now. Not entirely sure if we depend on something specific in the Mapsui color. 

The Color and Easing objects are there because they are static fields that hold a pointer to an instance. In case of color those are the 147 web colors and a transparent color. This number will not increase when creating more instances. Perhaps there are better solutions than what we are using, but this will not cumulate into a problem. I don't expect a page to stay in memory because of this.

> It's always the same resources
So, there are never two Map objects left?

You say you create a Map but in the screenshot it also shows a TileLayer, so you create a Map with a TileLayer, correct? If so, could you paste the code of what you are doing exactly?

The Map and its fields should be collected. We need to look into this.

**Comment by Omenrevlis** [comment]
This is exactly how I created the maps for the test. Yes I created a tile layer (as I would on a real page) and I crated another layer for pins as I would on a real page.

private void Button_Click(object sender, Microsoft.UI.Xaml.RoutedEventArgs e)
{
     _Maps = new List<Map>();
     for (var i = 0; i < 20; i++)
     {
         var map = new Map();
         map.Layers.Add(OpenStreetMap.CreateTileLayer());
         var layer = new GenericCollectionLayer<List<IFeature>>
         {
             Style = SymbolStyles.CreatePinStyle()
         };
         map.Layers.Add(layer);
         _Maps.Add(map);
     }
     txtNotice.Text= "Created!";
}

**Comment by Omenrevlis** [comment]
The test I performed was super simple. I didn't use Lambdas. I've attached the exact code I ran to test.  Super simple. Main page has two buttons and a textblock. The button events create and then dispose the objects.

The size of what remains is not the issue. The issue is these resources not disposed properly means the whole page stays in memory too.

Specifically, this is an issue in the Navigation template from WinUI Template Studio (Microsoft stuff). The way it works is there is a main page holds a navigation menu wired up for you and a <Frame> that your content pages will get loaded via a Helper class wired into the XAML created by the template. What happens when you use the navigation, it clears and disposes of the <Frames> content and loads a NEW instance of another page it's navigating to.

The problem comes when you navigate away from a page with Mapsui. It cannot release all its resources (mapsui) so the whole page stays in memory. If you navigate to the map page again, you get a new instance in memory. When you navigate to another page, the page stays in memory. If you keep going back and forth from map pages the memory use grows and grows. The zombie pages stay in memory until the application is closed.

In some applications this could be a massive issue. Imaging an application where you are delivering using a driving route, but also have to go to another screen to log hours. That application could stay open all the time and work for many drivers. Lots of other types of application could have issues.

**Comment by pauldendulk** [comment]
Thanks, I will try to reproduce it myself. 

Most of the memory in the screenshot can not cause the problem the problem you describe, (not the lambdas, not Colors or Easing, but there could be just one thing that is not disposed properly.

**Comment by Omenrevlis** [comment]
Here is a link to a sample of a navigation project. I created two pages. The main page has a map. The other page is blank.

https://drive.google.com/file/d/1rLAPvgw_YCjlpJ-G4AQaF18dGhRWxLMW/view?usp=sharing

I added the winui NuGet package.

This is with Visual Studio 2022
Windows Software Development Kit - Windows 10.0.22621.2428
Template Studio for WinUI (C#)  - https://marketplace.visualstudio.com/items?itemName=TemplateStudio.TemplateStudioForWinUICs

From: Mike Conley
Sent: Friday, November 10, 2023 11:31 AM
To: Mapsui/Mapsui ***@***.***>; Mapsui/Mapsui ***@***.***>
Cc: Author ***@***.***>
Subject: RE: [Mapsui/Mapsui] Memory leak WinUI 3 (Issue #2259)

Here is a memory snapshot of an actual Mapsui in a Page with 1 layer. I navigate away from it 25 times, then I waited a few minutes and took a snapshot. This is using the Navigation template as intended. I do not call the dispose, the page does that internally. This is a more accurate representation of what will happen in an application and not just a controlled test.

TLDR: 103 megabytes of just mapsui stuff in memory

**Comment by Omenrevlis** [comment]
I figure I should mention that going down the Idisposable route on the client application and figuring out where to call it doesn't work. I've tried. I can put that in the sample if you want, but essentially the control reference has been disposed before you can call to dispose it.

From: Mike Conley
Sent: Friday, November 10, 2023 11:57 AM
To: Mapsui/Mapsui ***@***.***>; Mapsui/Mapsui ***@***.***>
Cc: Author ***@***.***>
Subject: RE: [Mapsui/Mapsui] Memory leak WinUI 3 (Issue #2259)

Here is a link to a sample of a navigation project. I created two pages. The main page has a map. The other page is blank.

https://drive.google.com/file/d/1rLAPvgw_YCjlpJ-G4AQaF18dGhRWxLMW/view?usp=sharing

I added the winui NuGet package.

This is with Visual Studio 2022
Windows Software Development Kit - Windows 10.0.22621.2428
Template Studio for WinUI (C#)  - https://marketplace.visualstudio.com/items?itemName=TemplateStudio.TemplateStudioForWinUICs


From: Mike Conley
Sent: Friday, November 10, 2023 11:31 AM
To: Mapsui/Mapsui ***@***.******@***.***>>; Mapsui/Mapsui ***@***.******@***.***>>
Cc: Author ***@***.******@***.***>>
Subject: RE: [Mapsui/Mapsui] Memory leak WinUI 3 (Issue #2259)

Here is a memory snapshot of an actual Mapsui in a Page with 1 layer. I navigate away from it 25 times, then I waited a few minutes and took a snapshot. This is using the Navigation template as intended. I do not call the dispose, the page does that internally. This is a more accurate representation of what will happen in an application and not just a controlled test.

TLDR: 103 megabytes of just mapsui stuff in memory

**Comment by pauldendulk** [comment]
@Omenrevlis Did not get around to this one yet, but one question came up. Could you also test this when also calling Dispose on the layer? This  is actually necessary, we do not dispose the layers when disposing the map.

I just created a US to make an alteration https://github.com/Mapsui/Mapsui/issues/2284

**Comment by Omenrevlis** [comment]
Tested it. First, I would say it would be nice to have the map dispose it layers itself.
As for the test, similar results. Here is a screenshot of memory usage after adding 20 with 2 layers then looping through and disposing the layers before the maps.

***@***.***

Here is the simple code that I used.

private void Button_Click(object sender, Microsoft.UI.Xaml.RoutedEventArgs e)
{
     _Maps = new List<Map>();
     for (var i = 0; i < 20; i++)
     {
         var map = new Map();
         map.Layers.Add(OpenStreetMap.CreateTileLayer());
         var layer = new GenericCollectionLayer<List<IFeature>>
         {
             Style = SymbolStyles.CreatePinStyle()
         };
         map.Layers.Add(layer);
         _Maps.Add(map);
     }
     txtNotice.Text = "Created!";
}

private void Button_Click_1(object sender, Microsoft.UI.Xaml.RoutedEventArgs e)
{
     foreach (var map in _Maps)
     {
         foreach( var layer in map.Layers)
         {
             layer.Dispose();
         }
      map.Widgets.Clear();
map.ClearCache();
         map.Dispose();
     }
     _Maps.Clear();
     _Maps = null;
     GC.Collect();
     txtNotice.Text = "Disposed!";
}


Now for the real test, disposing in a page when navigating away from the page using the Navigation Template Studio solution.
For this I attached to the Page.Unloaded event and used the follow dispose code. I also put a destructor last ditch attempt for fun. Here is the code. I don't use widgets for this, but I figured why not clear it too.

private void Page_Unloaded(object sender, Microsoft.UI.Xaml.RoutedEventArgs e)
{
    Dispose();
}

public void Dispose()
{
    if(null != MyMap && null != MyMap.Map) {
        foreach (var layer in MyMap.Map.Layers)
        {
            layer.Dispose();
        }
        MyMap.Map.Widgets.Clear();
        MyMap.Map.ClearCache();
        MyMap.Map.Dispose();
    }
}

~MapPage()
{

    try { Dispose(); }
    catch
    { //do nothing this is last ditch attempt
    }
}

And the list is too big to lay out, so I attached a spreadsheet to the email.


On a side note,  perhaps you are willing to let me know two unrelated things? Is there a way to get a map with only English on it? Also, is there a way to get a map that is globe that can be spun?

Thank you,
Mike Conley
